### DISCUSSION TITLE
Google: Harmonize Google plugin auth

### DISCUSSION THREAD

**Discussion by sunker** [grafana/grafana#39907] [Data Sources]
**What would you like to be added?**:

I would like all Google plugins to have the same authentication capabilities and the code and documentation for this to be handled in one single place.

**Why?**

The authentication methods `JWT File` and `GCE Default Service Account` are currently only reusable for core plugins and for external plugins that doesn't have a backend.  Also, providing shared packages for handling Google plugin auth will make it easier for us and the community to build new Google plugins.

---

**Existing Google plugins**

Currently, Grafana maintain three Google plugins. The [CloudMonitoring datasource](https://github.com/grafana/grafana/blob/main/pkg/tsdb/cloudmonitoring/cloudmonitoring.go) which is a part of core Grafana plugin maintained by the cloud data sources squad, the [BigQuery plugin](https://github.com/grafana/google-bigquery-datasource) which is an external plugin maintained by the BI squad and the [google sheets plugin](https://github.com/grafana/google-sheets-datasource) plugin which is also an external plugin that is being maintained by the community squad. 

**Authentication methods**

***JWT Key File***
A Service Accounts can be used for accessing APIs in a certain GCP project. It's possible to generate a JWT key file that is associated with the service account. The credentials in this file can then be used to generate an OAuth 2.0 token that can be used in API calls to google APIs. 

The code for handling JWT key file auth is located in the [data source proxy](https://github.com/grafana/grafana/blob/main/pkg/api/pluginproxy/ds_auth_provider.go#L104-L109). This authentication method is currently available in CloudMonitoring and BigQuery. Google sheets also has support for JWT key file auth, however it's not using the auth proxy. The google sheets plugin uses the google go sdk for interacting with the service which is fine, however the token generated for the JWT file is not being cached so there's room for improvements here. 

***GCE Default Service Account***
Virtual machines in Google Compute Engine can associated with a Service Account. If Grafana is running on a GCE instance, Grafana can obtain an OAuth2 token for the Service Account using the [GCE meta data server api](https://cloud.google.com/compute/docs/metadata/overview). This token can then be used to access google APIs, given the service account has the associated permissions. 

The code for handling GCE auth is located in the [data source proxy](https://github.com/grafana/grafana/blob/main/pkg/api/pluginproxy/ds_auth_provider.go#L97-L102). This authentication method is currently available in CloudMonitoring and BigQuery, but should be a valid auth method also for Google Sheets (https://github.com/grafana/google-sheets-datasource/issues/144). 

***API Key***
This authentication option can be used for accessing public spreadsheets in Google Sheets. It's likely not a valid authentication method for other plugins, so this probably doesn't belong in a shared SDK. 

**The data source proxy**
The [data source proxy](https://grafana.com/docs/grafana/latest/developers/plugins/add-authentication-for-data-source-plugins/#authenticate-using-the-data-source-proxy) in core Grafana provides different methods of authenticating requests to a data source api by specifying routes in the plugin.json file. 

***Limitations of the data source proxy***
The initial idea of this proxy was to provide authentication to plugins that doesn't have a backend. These authentication methods might also be usable for plugins that has a backend - like in the QueryDataHandler for example. For core plugins, this is possible to achieve by using the [ApplyProxy](https://github.com/grafana/grafana/blob/main/pkg/tsdb/cloudmonitoring/cloudmonitoring.go#L551) function in the proxy. This function is however not available to external plugins, so currently there's no way of providing `JWT Key file auth` or `GCE auth` in the backend for these plugins. 

**Suggested solution**
The logic for providing tokens should not be in the plugin proxy, but in an go package in a separate repository. This go package should house logic for retrieving, caching and refreshing tokens for JWT file auth and GCE auth. It should expose methods for retrieving these tokens and take scopes as an argument. It should also expose an auth middleware (similar to [this one](https://github.com/grafana/grafana/blob/main/pkg/tsdb/azuremonitor/aztokenprovider/middleware.go)) that can be registered in the http clients for backend plugins. In order to not introduce breaking changes, the data source proxy should still support gce auth and jwt auth. However they should use token providers from the shared go package. 

The work can be split up into multiple tasks and separate issues can be created for most of them
- [x] The CloudMonitoring data source needs to be refactored to use the latest plugin-sdk api:s instead of the old tsdb api. This work has been started by the backend squad and I think it's close to being done. Please see [this](https://github.com/grafana/grafana/pull/38650) PR
- [x] Create new repo, `grafana-google-sdk` - DONE. See https://github.com/grafana/grafana-google-sdk-go
- [x] Move token provider logic from the authproxy in core Grafana to gce auth and jwt auth to the `grafana-google-sdk`. https://github.com/grafana/grafana/issues/39950
- [x] Use the new token providers exposed by the SDK in the [ds_auth_provider](https://github.com/grafana/grafana/blob/main/pkg/api/pluginproxy/ds_auth_provider.go) https://github.com/grafana/grafana/issues/40413
- [x] Refactor the CloudMonitoring backend so that instead of calling [ApplyRoute](https://github.com/grafana/grafana/blob/main/pkg/tsdb/cloudmonitoring/cloudmonitoring.go#L551) it shoud use the AuthMiddleware from the shared SDK. https://github.com/grafana/grafana/issues/40786
- [x] Refactor the CloudMonitoring plugin so that client side requests use the CallResourceHandler endpoint in the backend plugin API instead of the proxy. https://github.com/grafana/grafana/issues/40949
- [x] Refactor the CloudMonitoring plugin to process resource data on the backend and use CheckHealth instead of test. https://github.com/grafana/grafana/issues/41377
- [x] The way these authentication methods are configured in the data source configuration page look different in all Google plugins. That should not be the case, and to prevent that and make it easier build new plugins, we should keep configuration page logic inside an npm package, similar to what has been done in the [grafana/aws-sdk](https://github.com/grafana/grafana-aws-sdk/tree/main/src). https://github.com/grafana/grafana/issues/41764
- [x] Documentation for how to configure Google plugin auth should be kept in one place only, and all Google plugins should just reference this document in their documentation. Similar to what's been done for [AWS plugins](https://github.com/grafana/grafana/blob/main/docs/sources/datasources/aws-cloudwatch/aws-authentication.md). https://github.com/grafana/grafana/issues/41765

This work will be very similar to what has been done/is ongoing for Azure (see [this](https://github.com/grafana/grafana/issues/33809) epic) and for AWS. 

Feedback appreciated @grafana/cloud-datasources @grafana/grafana-bi-squad

**Comment by dprokop** [comment]
Thanks, @sunker for creating this, think this is a great proposal and will help a lot in maintaining google data sources. 

cc @ryantxu - something we talked about not that long ago ^^

**Comment by vickyyyyyyy** [comment]
Done in https://github.com/grafana/grafana/issues/44537
